tools/src/main/scala/scala/scalanative/codegen/Lower.scala

```diff
@@ -601,7 +603,11 @@ object Lower {
         else nir.MemoryOrder.Unordered
 
       // Acquire memory fence before loading a final field
-      if (field.attrs.isFinal) buf.fence(nir.MemoryOrder.Acquire)
+      // It's required only on the first access to obj fields
+      if (field.attrs.finalWithSafePublish && acquiredFinalFieldsOf.get.add(
+            obj
+          ))
+        buf.fence(nir.MemoryOrder.Acquire)
```

Comment:
What is this doing exactly? It's required on the first access _per thread_.

Reply:
It's a very simple optimisation, for cases when we want to load multiple fields from the same object. This way we have at most 1 fence per object within a single method. Previously we're introducing a fence for each load, although now as I'm thinking about it it's bugged and should be reverted - we might add fence in only one conditional branch of code, leaving other one vulnerable.

Reply:
Oops, good catch.

I think we should always emit NIR with all the barriers. Then we can implement an NIR => NIR optimization phase that attempts to "coalesce" multiple barriers. The JVM implements these sorts of optimizations.

I am trying to find more information about this, but here is an example:
https://bugs.openjdk.org/browse/JDK-8032218

Reply:
Yes, I was thinking about this, and also with coalescing of synchronized blocks. With these it should be easier, although probably we should create a dedicated NIR instruction for them to make it easier. Moving generation of fences from lowering to NIR generation would be required to correctly optimize them, but it would also harder to identify possibly redundant fences when using relaxed memory model - we'd need to perform a look ahead to check that. Anyway that's one important change to do